Platform: Allow copy of non copyable objects

pan- opened the pull request:
## Description

~~This patch adds the compile time configuration "non-copyable-warning" to mbed platform library. When this configuration flag is set to true, compile time errors resulting of the copy of a NonCopyable object are transformed into compile time and runtime warning.~~

~~This may be useful for *functional* legacy code that erroneously copy non copyable resources. It is highly encouraged to fix non complying code; copy of objects not designed to be copied may lead to resources leak, operation on invalid memory location as well as random malfunctioning.~~

This patch turns compile time errors resulting of the copy of a NonCopyable object into compile time and runtime warnings when the code is compiled with the develop or release profile. The compile time error is still issued when the code is compiled with the debug profile. 

Compile time error can be enforced by setting the compile time configuration "force-non-copyable-error" to true.

This patch also adds the macro `MBED_PRETTY_FUNCTION` which yield the string literal of the enclosing function signature. 

## Status

**READY**

## Migrations

NO

0xc0170 reviewed (changes requested):
Many style changes that distract from actual changes, can you please split the style changes into own separate commit?

pan- committed:
Platform: fix end of line whitespace

pan- committed:
Platform: Add MBED_PRETTY_FUNCTION macro.

This macro yields a string literal of the enclosing function name.

pan- committed:
Platform: Turn NonCopyable compile time error into warnings.

Turn the compile time error issued when a NonCopyable resource is copied
into a compile time and runtime warning.

If the application is compiled with the debug profile the compile time
error remains.

The compile time error can be enforced by setting the library option
force-non-copyable-error to true.
